modules/cmdrunner/widget.go

```diff
@@ -121,23 +123,45 @@ func runCommandLoop(widget *Widget) {
 		<-widget.runChan
 		widget.resetBuffer()
 		cmd := exec.Command(widget.settings.cmd, widget.settings.args...)
-		cmd.Stdout = widget
 		cmd.Env = widget.environment()
-		err := cmd.Run()
-
-		// The command has exited, print any error messages
+		var err error
+		if widget.settings.pty {
+			err = runCommandPty(widget, cmd)
+		} else {
+			err = runCommand(widget, cmd)
+		}
 		if err != nil {
-			widget.m.Lock()
-			_, writeErr := widget.buffer.WriteString(err.Error())
-			if writeErr != nil {
-				return
-			}
-			widget.m.Unlock()
+			widget.handleError(err)
 		}
 		widget.redrawChan <- true
 	}
 }
 
+func runCommand(widget *Widget, cmd *exec.Cmd) error {
+	cmd.Stdout = widget
+	return cmd.Run()
+}
+
+func runCommandPty(widget *Widget, cmd *exec.Cmd) error {
+	f, err := pty.Start(cmd)
+	// The command has exited, print any error messages
+	if err != nil {
+		return err
+	}
+
+	_, err = io.Copy(widget.buffer, f)
+	return err
+}
+
+func (widget *Widget) handleError(err error) {
+	widget.m.Lock()
+	_, writeErr := widget.buffer.WriteString(err.Error())
+	if writeErr != nil {
+		return
```

Comment:
Won't this leave the widget locked if WriteString fails?

Reply:
Yeah, like L76 this should probably be `defer widget.m.Unlock()`

Reply:
You are correct, issue with the original code. Will fix.